Use a global namecache to read `.hie` files

wz1000 opened the pull request:
This backports [this GHC MR](https://gitlab.haskell.org/ghc/ghc/-/merge_requests/3199) in preparation for a larger role for `.hie` files in ghcide.

We now have a global shared `NameCache` referenced by `ShakeExtras`, which we use to read `.hie` files.

cocreature commented on src-ghc810/Development/IDE/GHC/HieBin.hs:
```diff
@@ -0,0 +1,394 @@
+{-
```
I’m increasingly starting to wonder if GHC is the right location for the `.hie` files code. We seem to copy paste basically everything since we need to patch it. Hopefully, things will stabilize at some point :disappointed:

wz1000 replied:
It's used by haddock, so there isn't much of an alternative.

cocreature commented:
Needs rebasing.

wz1000 commented:
I found a bug introduced by this, fixed it and modified the tests to exercise this code path. See the commit message of https://github.com/digital-asset/ghcide/pull/677/commits/96d1f30ff84dc87d59118630d6b6dbfb for an explanation.

cocreature commented:
Thanks for the bugfix!
